pkg/register/component.go

```diff
@@ -0,0 +1,202 @@
+		return "", err
+	}
+
+	// Get Realm configuration from database
+	var realmConf dto.RealmConfiguration
+	realmConf, err = c.configDBModule.GetConfiguration(ctx, realmName)
+	if err != nil {
+		c.logger.Info(ctx, "msg", "Can't get realm configuration from database", "err", err.Error())
+		return "", err
+	}
+
+	// Get an OIDC token to be able to request Keycloak
+	var accessToken string
+	accessToken, err = c.tokenProvider.ProvideToken(ctx)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't get OIDC token", "err", err.Error())
+		return "", err
+	}
+
+	// Registering should be disallowed if an enabled user already exists with the same email
+	var kcUser *kc.UserRepresentation
+	kcUser, err = c.checkExistingUser(ctx, accessToken, user)
+	if err != nil {
+		return "", err
+	}
+
+	var userID string
+	if kcUser == nil {
+		var chars = []rune("0123456789")
+		for i := 0; i < 10; i++ {
+			var username = c.generateUsername(chars, 8)
+
+			// Create the user in Keycloak
+			kcUser = &kc.UserRepresentation{}
+			kcUser.Username = &username
+			user.UpdateUserRepresentation(kcUser)
+			userID, err = c.keycloakClient.CreateUser(accessToken, c.realm, c.realm, *kcUser)
+			// Create success: just have to get the userID and exist this loop
+			if err == nil {
+				var re = regexp.MustCompile(`(^.*/users/)`)
+				userID = re.ReplaceAllString(userID, "")
+				break
+			}
+			userID = ""
+			switch e := err.(type) {
+			case errorhandler.Error:
+				if e.Status == http.StatusConflict {
+					// Username already exists
+					continue
+				}
+			}
+			c.logger.Warn(ctx, "msg", "Failed to create user through Keycloak API", "err", err.Error())
+			return "", err
+		}
+		if userID == "" {
+			c.logger.Warn(ctx, "msg", "Can't generate unused username after multiple attempts")
+			return "", errorhandler.CreateInternalServerError("username.generation")
+		}
+	} else {
+		userID = *kcUser.Id
+		user.UpdateUserRepresentation(kcUser)
+		err = c.keycloakClient.UpdateUser(accessToken, c.realm, userID, *kcUser)
+		if err != nil {
+			c.logger.Warn(ctx, "msg", "Failed to update user through Keycloak API", "err", err.Error())
+			return "", err
+		}
+	}
+
+	// Store user in database
+	err = c.usersDBModule.StoreOrUpdateUser(ctx, c.realm, apiregister.DBUser{
+		UserID:               &userID,
+		BirthLocation:        user.BirthLocation,
+		IDDocumentType:       user.IDDocumentType,
+		IDDocumentNumber:     user.IDDocumentNumber,
+		IDDocumentExpiration: user.IDDocumentExpiration,
+	})
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't store user details in database", "err", err.Error())
+		return "", err
+	}
+
+	// Send execute actions email
+	if realmConf.RegisterExecuteActions != nil && len(*realmConf.RegisterExecuteActions) > 0 {
+		err = c.keycloakClient.ExecuteActionsEmail(accessToken, c.realm, userID, *realmConf.RegisterExecuteActions)
+		if err != nil {
+			c.logger.Warn(ctx, "msg", "ExecuteActionsEmail failed", "err", err.Error())
```

Comment:
It would be good to log the userID and the list of registerActions, thus we would be able to do it manually later on.

Reply:
User will have to retry its account creation if executeActionsEmail fails